reasoner/Reasoner.java

```diff
@@ -120,15 +107,33 @@ private boolean hasRule(Label type) {
     }
 
     public FunctionalIterator<ConceptMap> execute(Disjunction disjunction, GraqlMatch.Modifiers modifiers, Context.Query context) {
-        resolveTypes(disjunction, list());
-        Set<Identifier.Variable.Name> filter =
-                iterate(modifiers.filter()).map(v -> v.reference().asName()).map(Identifier.Variable::of).toSet();
-        iterate(disjunction.conjunctions()).filter(c -> !c.isSatisfiable() && !isSchemaQuery(c, filter)).map(c -> {
-            throw GraknException.of(UNSATISFIABLE_CONJUNCTION, c);
-        });
+        logicMgr.typeResolver().resolve(disjunction);
+
+        if (!disjunction.isCoherent()) {
+            Set<Conjunction> causes = incoherentConjunctions(disjunction);
+            throw GraknException.of(UNSATISFIABLE_PATTERN, disjunction, causes);
+        }
 
         if (mayReason(disjunction, context)) return executeReasoner(disjunction, modifiers, context);
-        else return executeTraversal(disjunction, context, filter);
+        else return executeTraversal(disjunction, context, filter(modifiers.filter()));
+    }
+
+    private Set<Identifier.Variable.Name> filter(List<UnboundVariable> graqlVars) {
+        return iterate(graqlVars).map(v -> v.reference().asName()).map(Identifier.Variable::of).toSet();
+    }
+
+    private Set<Conjunction> incoherentConjunctions(Disjunction disjunction) {
+        assert !disjunction.isCoherent();
+        Set<Conjunction> causes = new HashSet<>();
+        for (Conjunction conjunction : disjunction.conjunctions()) {
+            FunctionalIterator<Negation> incoherentChildren = iterate(conjunction.negations()).filter(negation -> !negation.isCoherent());
+            if (!conjunction.isCoherent() && !incoherentChildren.hasNext()) {
+                causes.add(conjunction);
+            } else {
+                incoherentChildren.forEachRemaining(negation -> causes.addAll(incoherentConjunctions(negation.disjunction())));
+            }
```

Comment:
I can do that after we merge, but in the future note that if and else block can be made one line each without curly brackets

Reply:
i tried, the second line doesn't fit :(